### ISSUE TITLE
Show colors when using word diff

### ISSUE THREAD

**Issue by longouyang** [magit/magit#1615]
When editing LaTeX files, I tend to put entire paragraphs on a single line. If I make some change within the paragraph, the resulting diff is hard to spot. I've tried using using git's (colored) word-diff mode to make this easier by doing `(setq magit-diff-options (list "--word-diff=color"))` in my init file. This works, except it doesn't show the colors. I've tried setting the git global `color.diff` to always, true, and false, but none of these seem to work.

It looks like this has been brought up before ([2012 google groups thread](https://groups.google.com/forum/#!searchin/magit/%22--color-words%22/magit/yICvS2DJ-cw/BHflNl4K9r0J), [2010 google groups thread](https://groups.google.com/forum/#!searchin/magit/%22--color-words%22/magit/9BDhhfS3ZT4/aoQyTnz961oJ)) but no one has come up with a solution. I'm not good at elisp, but I'm willing to give this a shot. In principle, how would I make this work?

**Comment by tarsius** [comment]
One big problem with this is that it won't be possible to apply (stage...) the resulting hunks. Maybe that could be hacked around by applying the respective line-wise hunk, but even then it would not be possible to apply the region.

**Comment by tarsius** [comment]
Following up on what I said in my last comment:

... and that means I won't implement this. While "hunk refinement" doesn't do exactly the same, and might be less convenient as long as we only want to *look at* the differences, it has the considerable advantage that it does in not affect diff parsing (we basically just call a function which does the refinement) and applying a refined diff also just works as for a non-refined diff.

For git's word diff is not the case. Just implementing the parsing itself would not be all that hard, but the resulting diff would not be "actionable". Applying wouldn't work and many other commands that can act on other diffs would also not work. Some of them would do so gracefully, while others would fail with some arbitrary sounding error (e.g. a type error). But Magit users expect the thing at point to be actionable - especially diffs. So adding barely functional word diff would result in bug reports and feature requests puring in. So I am not adding word diff support - not even read-only support.

**Comment by jdtsmith** [comment]
Revisiting this ancient issue for the same reason (very long lines in LaTeX file making small diff's hard to spot).  I wonder if more/less context could be expanded to _narrow_ regions to the (near) minimum-set of words that include all changes before and after.  That way the text would still be there and could be applied, but those who want to focus in on just the part of the (very long) line which actually changed could do so as well.

**Comment by tarsius** [comment]
I don't understand.

**Comment by jdtsmith** [comment]
Old:
```diff
modified   main.tex
@@ -211,7 +211,7 @@ Blah blah

- This is a very long line with several sentences.  It is in fact a standalone paragraph.  When you create a small change within this "line", git considers the entire line to have changed.  Very sad.  But when only [one] word is changed, it is very inconvenient if the entire text of the line/aka paragraph is displayed.  Especially when it has several dozen sentences.  Writing several dozen sentences would be uninteresting, so just pretend this paragraph is much, much longer.  
+ This is a very long line with several sentences.  It is in fact a standalone paragraph.  When you create a small change within this "line", git considers the entire line to have changed.  Very sad.  But when only [a single] word is changed, it is very inconvenient if the entire text of the line/aka paragraph is displayed.  Especially when it has several dozen sentences.  Writing several dozen sentences would be uninteresting, so just pretend this paragraph is much, much longer.  
```


New (after invoking `-`, show less context, a "final" time, aka a _just the surrounding words_ context):

```diff
modified   main.tex
@@ -211,7 +211,7 @@ Blah blah

- ... when only [one] word is ...  
+ ... when only [a single] word is ...  
```

Thinking about this more, maybe magit could use the `invisible` [text property](https://www.gnu.org/software/emacs/manual/html_node/elisp/Invisible-Text.html) instead of narrowing.  Such that the full text of the diff hunk is still there (and can be applied). And if magit marks parts of a hunk with this special invisible property, it can toggle invisibility on and off as appropriate given the amount of context, using `buffer-invisibility-spec`.  Something like:

```elisp
(defun context-make-invisible (start end)
  (interactive "r")
  (set-text-properties start end '(invisible context-invisible)))
;; Select a region and M-x context-make-invisible
;; then evaluate these, in turn:
(setq buffer-invisibility-spec '((context-invisible . t)))
(setq buffer-invisibility-spec nil)
```

Then it would just be a matter of marking with the right `invisible` property all text in the hunk outside of the few words surrounding the changed bits, which magit is already highlighting (if it's diff refining).   Maybe even add a click-to-expand property on the ellipses, and perhaps give some control of how many surrounding words (sentences?) of context the user wants.  Crazy?

**Comment by tarsius** [comment]
If you implement it, then I would consider adding it here. (Though I might also end up preferring a separate package.)

**Comment by jdtsmith** [comment]
OK, thanks. Is there a hook should investigate which would be appropriate for diff styling?

**Comment by tarsius** [comment]
This is similar to "hunk refinement", which is being done by `magit-diff-update-hunk-refinement`. Look at how that is called.

**Comment by jdtsmith** [comment]
OK, I've cobbled something together as a test [here](https://github.com/jdtsmith/evenless-context/blob/master/evenless-context.el).  In fact I found that Magit delegates to `diff-mode`'s `diff-refine-hunk`, which further delegates refining to `smerge-refine-regions`, so I advise that.  Already I find it very useful for looking at diffs of long lines.  

I wonder if you could give it a brief try by `emacs-lisp-byte-compile-and-load`'ing that file.  I haven't tested with other diff formats like context-diff, nor have I attempted any of the apply-magic you mention, so I want to be sure this preserves the basic functionality of diff buffers before moving further.  If you could have a quick look I'd appreciate.  Hidden status can be changed in the diff buffer via `evenless-context-enable|disable-hiding`, and it works best with `refine` as `'all`.

I think a package would work fine here, though I note Magit already uses `'invisibility` to hide trailing CR's in diff's, and it does so using `t` as the property value, which doesn't play nicely with selective invisibility.  An easy fix. Haven't thought of an interface yet, but perhaps just a new toggle option in the `D` menu.

**Comment by tarsius** [comment]
It doesn't work for me. After enabling I can no longer toggle the visibility of any section.

> I think a package would work fine here, though I note Magit already uses `'invisibility` to hide trailing CR's in diff's, and it does so using `t` as the property value, which doesn't play nicely with selective invisibility. An easy fix.

And the easy fix would be?

Magit also uses that property to hide sections. This appears to be trampling on that.

**Comment by jdtsmith** [comment]
Thanks for trying this.  Not sure why I missed that.  I see the issue, which only occurs with abridging "off".  

The fix appears to be never using `t` for invisibility or the invisibility-spec, as 

The problem is that an `'invisibility` value of `t` is a heavy hammer, and sets or clobbers all invisibility in the buffer.  I.e. `t` matches any non-nil invisibility spec, and any invisibility property value matches `t`.   

The fix is not to use `t`. Instead, you could use `(add-to-invisibility-spec 'magit-section-invisible)`, and mark your sections using `'magit-section-invisible` or similar as the value instead of `t`.  Then magit can toggle section invisibility (either by altering the spec, or by manipulating overlays), and so can (separately) a tool like `abridge-diff`.  Apparently updating the invisibility spec is [much faster](https://www.gnu.org/software/emacs/manual/html_node/elisp/Invisible-Text.html) than searching for and (re-)moving overlays: 

> Setting this variable is very fast, much faster than scanning all the text in the buffer looking for properties to change.

**Comment by tarsius** [comment]
I am on vacation and don't want to test other peoples packages right now.

> The fix is not to use `t`. Instead, you could use `(add-to-invisibility-spec 'magit-section-invisible)`,

If you implement and thoroughly test that, then I would look at it. Right now I would not only have to re-read lots of documentation but also figure out again where the t is hardcoded in Magit, which I am also not thrilled about.

**Comment by jdtsmith** [comment]
**TLDR — No Magit changes are needed at the moment, except for future-proofing against further simultaneous uses of `'invisibility` by other associated packages, or Magit itself.**  

> I am on vacation and don't want to test other peoples packages right now.

Understood.  Enjoy. 
 
> > The fix is not to use `t`. Instead, you could use `(add-to-invisibility-spec 'magit-section-invisible)`,
> 
> If you implement and thoroughly test that, then I would look at it. Right now I would not only have to re-read lots of documentation but also figure out again where the t is hardcoded in Magit, which I am also not thrilled about.

OK, I can work around this, in fact already have.  Right now my package works by leaving the token `t` for invisibility as the sole domain of Magit.  The only issue remaining is that the default value for `buffer-invisibility-spec` is the _symbol_ `t`. The first call to `add-to-invisibility-spec` alters this to be a list containing `t`, and it remains a list thereafter (and hence fixes the bug you noticed).  In my plugin, this happens automatically via `'magit-diff-mode-hook` and `'magit-status-mode-hook` (hopefully those are the only places that diff-mode functionality is used in Magit).  

Thanks for your help.

**Comment by tarsius** [comment]
TLDR

Thanks :grinning:

**Comment by jdtsmith** [comment]
Hi @tarsius, I'd like to add an option to disable long diff abridging in magit on hunks from certain files matching a configurable list of patterns (*.el, *.py, etc.).  Is there a standard way to read the filename corresponding to a given hunk (wherever it may appear)?  I see the text property on the `@@` line like `#s(magit-file-section file #("mlscroll.el" 0 11 ...)` so could likely pull it out of there, but suspect there is a cleaner way.  

BTW, here's what `abridge-diff` adds to `D` (suffix `a`): 

![image](https://user-images.githubusercontent.com/93749/118588253-013c1d80-b76c-11eb-9b27-f067b6a56d1d.png)

**Comment by tarsius** [comment]
Use `magit-file-at-point` or maybe `magit-current-file`. See their definitions to decide which.
